pkg/authentication/oidc.go

```diff
@@ -119,6 +133,110 @@ func (o *OAuth2Context) RefreshToken(ctx context.Context, rawRefreshToken string
 		Token()
 }
 
+func (o *OAuth2Context) Logout(ctx context.Context, principal *Principal) error {
+	_, span := o.tracer.Start(ctx, "authentication.OAuth2Context.Logout")
+	defer span.End()
+
+	if principal == nil {
+		return fmt.Errorf("no principal provided")
+	}
+
+	httpClient := HTTPClientFromContext(ctx)
+
+	err := o.revokeSession(principal, httpClient)
+	if err != nil {
+		return err
+	}
+
+	// perform token revocation
+	return o.revokeToken(principal, httpClient)
+}
+
+func (o *OAuth2Context) revokeSession(principal *Principal, client HTTPClientInterface) error {
+	// in case of a CLI user no SessionID is present
+	if principal.SessionID == "" {
+		return nil
+	}
+
+	req, err := o.prepareSessionRevocationRequest(principal.SessionID)
+	if err != nil {
+		return err
+	}
+
+	return o.performRequest(client, req, http.StatusNoContent, "revoke session")
+}
+
+func (o *OAuth2Context) revokeToken(principal *Principal, client HTTPClientInterface) error {
+	token := principal.RawAccessToken
+	if principal.RawRefreshToken != "" {
+		token = principal.RawRefreshToken
+	}
+
+	req, err := o.prepareTokenRevocationRequest(token)
+	if err != nil {
+		return err
+	}
+
+	return o.performRequest(client, req, http.StatusOK, "revoke token")
+}
+
+func (o *OAuth2Context) performRequest(httpClient HTTPClientInterface, req *http.Request, okStatus int, requestName string) error {
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != okStatus {
+		return o.hydraApiError(requestName, resp)
+	}
+
+	return nil
+}
+
+func (o *OAuth2Context) hydraApiError(requestName string, resp *http.Response) error {
+	revokeErr := new(hydraAPIError)
+	defer resp.Body.Close()
+
+	if err := json.NewDecoder(resp.Body).Decode(revokeErr); err != nil {
+		return err
+	}
+
+	return fmt.Errorf("%s request failed, error: %s, description: %s", requestName, revokeErr.Error, revokeErr.ErrorDescription)
+}
+
+func (o *OAuth2Context) prepareSessionRevocationRequest(sessionID string) (*http.Request, error) {
+	req, err := http.NewRequest(http.MethodDelete, fmt.Sprintf("%s/admin/oauth2/auth/sessions/login", o.hydraAdminURL), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	query := req.URL.Query()
+	query.Add("sid", sessionID)
+
+	req.URL.RawQuery = query.Encode()
+	return req, err
+}
+
+func (o *OAuth2Context) prepareTokenRevocationRequest(token string) (*http.Request, error) {
+	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("%s/oauth2/revoke", o.hydraPublicURL), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+
+	err = json.NewEncoder(&buf).Encode(&revokeTokenRequest{Token: token})
+	if err != nil {
+		return nil, err
+	}
+
+	req.SetBasicAuth(o.client.ClientID, o.client.ClientSecret)
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Body = io.NopCloser(&buf)
+
+	return req, nil
```

Comment:
why are we not using the hydra sdk for this?

Reply:
Clarified offline, will proceed and refactor implementation and tests. Thanks

Reply:
Done